DeeperFunc/Deeper.swift

```diff
@@ -0,0 +1,75 @@
+//
+//  Deeper.swift
+//  DeeperFunc
+//
+//  Created by Ilya Puchka on 26/10/2017.
+//  Copyright © 2017 Ilya Puchka. All rights reserved.
+//
+
+import Foundation
+
+public enum Either<A, B> {
+    case left(A), right(B)
+}
+
+public typealias RouteComponents = (path: [String], query: [String: String])
+
+public protocol PatternState {}
+public protocol ClosedPatternState: PatternState {} // pattern is complete
+public protocol OpenPatternState: PatternState {} // pattern requires subsequent pattern to become closed
+
+public protocol ClosedPathPatternState: ClosedPatternState {}
+public enum Path: ClosedPathPatternState {} // somewhere in the path
+public enum Query: ClosedPatternState {} // query started
+
+public struct RoutePattern<A/*pattern type*/, S: PatternState> {
+    public let parse: Parser<A> // parses path components exctracting underlying type of pattern
+    public let print: Printer<A> // converts pattern with passed in value to template component
+    public let template: String
+    
+    func map<S>(_ iso: PartialIso<A, Any>) -> RoutePattern<Any, S> {
+        return .init(parse: {
+            guard let result = self.parse($0) else { return nil }
```

Comment:
not applying `iso.apply` here as then it will i.e. convert parsed `int` to `string` because it's the type of "type-erased" pattern wrapper used for parsing string format, and then router will not be able to cast in `map`. The only option is to use `ExpressibleByStringLiteral` or similar protocol to convert it back from string to type we need in router...